### DISCUSSION TITLE
[Question] How is memory calculated?

### DISCUSSION THREAD

**Discussion by roughnecks** [henrygd/beszel#80] [Support]
Hi there,

how comes memory usage (as reported by Beszel) is quite different from other apps like btop or the classic "free" command?

At the moment Beszel reports 42.0% RAM usage, while this tells me:

```
$ free -m
               total        used        free      shared  buff/cache   available
Mem:            7953        3941        3397         261        1191        4011
Swap:           8191           1        8190
```

which is about 49% used.
I never **really** understood all the values reported by `free`, eheh, but something doesn't look right to me here.

**Comment by henrygd** [comment]
I've never compared directly and not on a computer currently but the beszel number is meant to be memory used by programs whereas the btop / free calculation probably includes slab or some other reclaimable memory.

I'm using the 'Used' calculation from the gopsutil library, which reads from /proc/meminfo. For more info see that lib.

If people think memory should be more in line with free then it should be possible to do that. Maybe as an env var option for what calculation to use.

**Comment by roughnecks** [comment]
Just found this: https://github.com/shirou/gopsutil/issues/899

**Comment by henrygd** [comment]
I don't think there's a single value that's objectively correct when it comes to Linux. People have different opinions on it.

For example, it looks like htop displays a value closer to our number than free / btop.

I use [Astra Monitor](https://github.com/AstraExt/astra-monitor) on my laptop and calculation they use by default is the same or very close to ours currently.

I'm not opposed to tweaking so it's more in line with free though.

**Comment by roughnecks** [comment]
Sorry to bother you, but I've got another perplexity about measurements.
Every morning at 5 AM my VPS does backup some data (1GB or so, not very much) and I'm seeing these two graphs:

![image](https://github.com/user-attachments/assets/c035da73-7eef-4236-a495-e40da387a571)

![image](https://github.com/user-attachments/assets/97ac9c97-242b-4664-847f-0879698dd729)

CPU may be alright, but disk read and write looks too low for me. VPS has got a SSD and I'm quite sure it performs faster that 2MB/s ..

Obviously I'm not able to check this in real time, since I'm quite used to sleep :)
Any suggestions?

Thanks

**Comment by henrygd** [comment]
The value is the average over the course of the time block in question.

So in the 12 hour chart, 1.8 MB/s should be what was averaged over the course of the whole 10 minute period. Not the peak value.

If your backup is 1 GB, divide by 600 seconds and you get 1.66 MB/s.

I can look into adding peak value at some point.

**Comment by henrygd** [comment]
I converted this to a discussion, since it's not really an issue and maybe someone else will have an opinion on whether we should align with free.

**Comment by roughnecks** [comment]
cool
does this looks normal to you?

https://i.imgur.com/ceNGHtm.jpeg

**Reply by henrygd** [reply]
What about it? The y axis ticks? Might be better to have a 6 GB line, but that's the way the lib scales it by default. I think it uses d3-scale.

I did some testing a while back by overriding the tick number to 9, which seemed to avoid those gaps most of the time, but not sure if it ended up doing something weird with different amounts of total ram.

**Reply by roughnecks** [reply]
yeah, y axis :)
don't worry, I was just curious.
